design/plugin-backup-and-restore-progress-design.md

```diff
@@ -0,0 +1,238 @@
+- not knowing the backup information during retore operation
+- need to be optional as few plugins may not have a way to provide the progress information
+
+This document proposes an approach for plugins to follow to provide backup/restore progress, which can be used by users to know the progress.
+
+## Goals
+
+- Provide uniform way of visibility into backup/restore operations performed by volume snapshotters
+
+## Non Goals
+- Plugin implementation for this approach
+
+## Background
+
+(Omitted, see introduction)
+
+## High-Level Design
+
+### Progress of backup operation handled by volume snapshotter
+
+Progress will be updated by volume snapshotter in VolumePluginBackup CR which is specific to that backup operation.
+
+### Progress of restic operation handled by volume snapshotter
+
+Progress will be updated by volume snapshotter in VolumePluginRestore CR which is specific to that restore operation.
+
+## Detailed Design
+
+### VolumePluginBackup CR
+
+```
+// VolumePluginBackupSpec is the specification for a VolumePluginBackup CR.
+type VolumePluginBackupSpec struct {
+	// Volume is the PV name to be backed up.
+	Volume string `json:"volume"`
+
+	// Backup name
+	Backup string `json:"backup"`
+
+	// Plugin name
+	Plugin string `json:"plugin"`
+
+	// PluginSpecific are a map of key-value pairs that plugin want to provide
+	// to user to identify plugin properties related to this backup
+	// +optional
+	PluginSpecific map[string]string `json:"pluginSpecific,omitempty"`
+}
+
+// VolumePluginBackupStatus is the current status of a VolumePluginBackup CR.
+type VolumePluginBackupStatus struct {
+	// Phase is the current state of the VolumePluginBackup.
+	Phase string `json:"phase"`
+
+	// SnapshotID is the identifier for the snapshot of the volume.
+	// This will be used to relate with output in 'velero describe backup'
+	SnapshotID string `json:"snapshotID"`
+
+	// Message is a message about the volume plugin's backup's status.
+	// +optional
+	Message string `json:"message,omitempty"`
+
+	// StartTimestamp records the time a backup was started.
+	// Separate from CreationTimestamp, since that value changes
+	// on restores.
+	// The server's time is used for StartTimestamps
+	// +optional
+	// +nullable
+	StartTimestamp *metav1.Time `json:"startTimestamp,omitempty"`
+
+	// CompletionTimestamp records the time a backup was completed.
+	// Completion time is recorded even on failed backups.
+	// Completion time is recorded before uploading the backup object.
+	// The server's time is used for CompletionTimestamps
+	// +optional
+	// +nullable
+	CompletionTimestamp *metav1.Time `json:"completionTimestamp,omitempty"`
+
+	// Progress holds the total number of bytes of the volume and the current
+	// number of backed up bytes. This can be used to display progress information
+	// about the backup operation.
+	// +optional
+	Progress VolumeOperationProgress `json:"progress,omitempty"`
+}
+
+type VolumeOperationProgress struct {
```

Comment:
This progress struct makes it seem like the snapshotter is performing a byte-by-byte copy of the data in the volume. Do we know which volume snapshotter plugins operate this way and which amongst them offer fine level of detail into their volume snapshot progress?

Reply:
lol. I would expect that too!
The specific question I was implying, is this something that is specific to openEBS?
How would other plugins be able to make use of these CRDs?

Reply:
I picked this Progress struct from PodVolumeBackup CR (https://github.com/vmware-tanzu/velero/blob/master/pkg/apis/velero/v1/pod_volume_operation_progress.go).
and OpenEBS, being block device provider, provides byte-level granularity.

Reply:
>  How would other plugins be able to make use of these CRDs?
This should be able to be used by any plugin by filling details of the amount of data that it need to backup and amount of data that it already backed up.
Existing restic operator in velero also fills similar details.

Reply:
Also, looks like there is quite a bit of overlap between the existing [pod+volume_backup](https://github.com/vmware-tanzu/velero/blob/master/pkg/apis/velero/v1/pod_volume_backup.go) CRD and what is in this proposal.
Is there a reason why we might not be able to re-use/extend the existing pod volume backup CRDs?

Reply:
PodVolumeBackups are currently only used for volumes protected by restic. They are not created at all for snapshots created via plugins.

I'm a little wary of overloading them too much, but not completely averse to the idea, depending on the details.

Reply:
I think using PodVolumeBackups would require modifying the controller at https://github.com/vmware-tanzu/velero/blob/master/pkg/controller/pod_volume_backup_controller.go to ignore any being created for use with plugins. We could do this, but I don't know exactly what the changes necessary would be at the moment. Similar goes for the restore side.

Reply:
marking it as resolved as per above comment.. please reopen if more info is required